time-aware-cross-validation/script.whizzml

```diff
@@ -0,0 +1,98 @@
+(define (safe-delete id exclude)
+  (when (not (member? id exclude))
+    (try (delete id)
+         (catch e
+           (log-info (str "Error deleting resource " id " ignored"))
+           false))))
+
+;; We divide all the dataset rows into (n-splits+1) intervals.
+;; This returns the position of the first row of interval i,
+;; using zero-based indexing
+(define (time-interval rows n-splits i)
+  (round (* (/ i (+ 1 n-splits)) rows)))
+
+;; Defines a range of intervals and returns a list with its start
+;; and end row positions. We add 1 to the start position because
+;; we don't want ranges to overlap. Furthermore, Bigml's
+;; rows ranges are closed intervals and use one-based indexing
+(define (interval-range rows n-splits from to)
+  (list (+ 1 (time-interval rows n-splits from))
+        (time-interval rows n-splits to)))
+
+(define (calculate-splits-expanding-window rows n-splits)
+  (log-info "Calculating expanding window splits...")
+  (map (lambda (i) (list (interval-range rows n-splits 0 (+ i 1))
+                         (interval-range rows n-splits (+ i 1) (+ i 2))))
+       (range n-splits)))
+
+(define (calculate-splits-rolling-window rows n-splits)
+  (log-info "Calculating rolling window splits...")
+  (map (lambda (i) (list (interval-range rows n-splits i (+ i 1))
+                         (interval-range rows n-splits (+ i 1) (+ i 2))))
+       (range n-splits)))
+
+;; Returns a list where each item (representing an split) contains
+;; the start and end positions of train and test rows intervals
+(define (calculate-splits rows n-splits fixed-window-size?)
+  (if fixed-window-size?
+    (calculate-splits-rolling-window rows n-splits)
+    (calculate-splits-expanding-window rows n-splits)))
+
+(define (create-cv-datasets ds-id name splits i)
```

Comment:
I'm not sure as to the method you used to compute the windows [from to] intervals. Maybe it works, but under some restrictions like rows > n-splits?

``` 
whizzml> (calculate-splits-expanding-window 3 3)
LOG info: Calculating expanding window splits...
[[[1 1] [2 2]] [[1 2] [3 2]] [[1 2] [3 3]]]
```

Reply:
And should the size of the training dataset and the test dataset be equal?

Reply:
Yes, I am assuming that the number of total rows is much bigger than the number of splits. In that case, I think everything is working. In the readme images you can find the expected response in a specific case, which is the same as the obtained in the repl:

`whizzml> (calculate-splits-expanding-window 24 5)
LOG info: Calculating expanding window splits...
[[[1 4] [5 8]]
 [[1 8] [9 12]]
 [[1 12] [13 16]]
 [[1 16] [17 20]]
 [[1 20] [21 24]]]`

`whizzml> (calculate-splits-rolling-window 24 5)
LOG info: Calculating rolling window splits...
[[[1 4] [5 8]]
 [[5 8] [9 12]]
 [[9 12] [13 16]]
 [[13 16] [17 20]]
 [[17 20] [21 24]]]`

In the rolling-window case training datasets have the same length as test ones. Some of the references I used did in that way, but I don't think it's mandatory, we could create bigger training datasets and smaller test ones.

Reply:
With the new commit I just pushed the maximum allowed value for **n-splits** is set to `0.1*rows` to avoid this cofussion